precompiles/balances-erc20/src/lib.rs

```diff
@@ -0,0 +1,413 @@
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+/// Solidity selector of the Transfer log, which is the Keccak of the Log signature.
+pub const SELECTOR_LOG_TRANSFER: &[u8; 32] =
+	u8_slice!("0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef");
+
+/// Solidity selector of the Approval log, which is the Keccak of the Log signature.
+pub const SELECTOR_LOG_APPROVAL: &[u8; 32] =
+	u8_slice!("0x8c5be1e5ebec7d5bd14f71427d1e84f3dd0314c0f7b2291e5b200ac8c7c3b925");
+
+/// Associates pallet Instance to a prefix used for the Approves storage.
+/// This trait is implemented for () and the 16 substrate Instance.
+pub trait InstanceToPrefix {
+	/// Prefix used for the Approves storage.
+	type ApprovesPrefix: StorageInstance;
+}
+
+// We use a macro to implement the trait for () and the 16 substrate Instance.
+macro_rules! impl_prefix {
+	($prefix:ident, $instance:ty, $name:literal) => {
+		pub struct $prefix;
+
+		impl StorageInstance for $prefix {
+			const STORAGE_PREFIX: &'static str = "Approves";
+
+			fn pallet_prefix() -> &'static str {
+				$name
+			}
+		}
+
+		impl InstanceToPrefix for $instance {
+			type ApprovesPrefix = $prefix;
+		}
+	};
+}
+
+impl_prefix!(ApprovesPrefix0, (), "Erc20Instance0Balances");
+impl_prefix!(ApprovesPrefix1, Instance1, "Erc20Instance1Balances");
+impl_prefix!(ApprovesPrefix2, Instance2, "Erc20Instance2Balances");
+impl_prefix!(ApprovesPrefix3, Instance3, "Erc20Instance3Balances");
+impl_prefix!(ApprovesPrefix4, Instance4, "Erc20Instance4Balances");
+impl_prefix!(ApprovesPrefix5, Instance5, "Erc20Instance5Balances");
+impl_prefix!(ApprovesPrefix6, Instance6, "Erc20Instance6Balances");
+impl_prefix!(ApprovesPrefix7, Instance7, "Erc20Instance7Balances");
+impl_prefix!(ApprovesPrefix8, Instance8, "Erc20Instance8Balances");
+impl_prefix!(ApprovesPrefix9, Instance9, "Erc20Instance9Balances");
+impl_prefix!(ApprovesPrefix10, Instance10, "Erc20Instance10Balances");
+impl_prefix!(ApprovesPrefix11, Instance11, "Erc20Instance11Balances");
+impl_prefix!(ApprovesPrefix12, Instance12, "Erc20Instance12Balances");
+impl_prefix!(ApprovesPrefix13, Instance13, "Erc20Instance13Balances");
+impl_prefix!(ApprovesPrefix14, Instance14, "Erc20Instance14Balances");
+impl_prefix!(ApprovesPrefix15, Instance15, "Erc20Instance15Balances");
+impl_prefix!(ApprovesPrefix16, Instance16, "Erc20Instance16Balances");
+
+/// Alias for the Balance type for the provided Runtime and Instance.
+pub type BalanceOf<Runtime, Instance = ()> =
+	<Runtime as pallet_balances::Config<Instance>>::Balance;
+
+/// Storage type used to store approvals, since `pallet_balances` doesn't
+/// handle this behavior.
+/// (Owner => Allowed => Amount)
+pub type ApprovesStorage<Runtime, Instance> = StorageDoubleMap<
+	<Instance as InstanceToPrefix>::ApprovesPrefix,
+	Blake2_128Concat,
+	<Runtime as frame_system::Config>::AccountId,
+	Blake2_128Concat,
+	<Runtime as frame_system::Config>::AccountId,
+	<Runtime as pallet_balances::Config<Instance>>::Balance,
```

Comment:
Why not use `BalanceOf<Runtime>`?

Reply:
`BalanceOf<Runtime, Instance>` indeed like used elsewhere in this file :)